File: framework/frontend/flutter/packages/k1s0_realtime/test/websocket/backoff_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:k1s0_realtime/src/types/reconnect_config.dart';
import 'package:k1s0_realtime/src/utils/backoff.dart';

void main() {
  group('calculateBackoff', () {
    group('linear backoff', () {
      test('試行回数に比例して遅延が増加する', () {
        expect(
          calculateBackoff(0, const Duration(seconds: 1),
              const Duration(seconds: 30), BackoffType.linear),
          const Duration(seconds: 1),
        );
        expect(
          calculateBackoff(1, const Duration(seconds: 1),
              const Duration(seconds: 30), BackoffType.linear),
          const Duration(seconds: 2),
        );
        expect(
          calculateBackoff(2, const Duration(seconds: 1),
              const Duration(seconds: 30), BackoffType.linear),
          const Duration(seconds: 3),
        );
      });

      test('最大遅延を超えない', () {
        expect(
          calculateBackoff(100, const Duration(seconds: 1),
              const Duration(seconds: 30), BackoffType.linear),
          const Duration(seconds: 30),
        );
      });
    });

    group('exponential backoff', () {
      test('試行回数の指数関数で遅延が増加する', () {
        expect(
          calculateBackoff(0, const Duration(seconds: 1),
              const Duration(seconds: 60), BackoffType.exponential),
          const Duration(seconds: 1),
        );
        expect(
          calculateBackoff(1, const Duration(seconds: 1),
              const Duration(seconds: 60), BackoffType.exponential),
          const Duration(seconds: 2),
        );
        expect(
          calculateBackoff(2, const Duration(seconds: 1),
              const Duration(seconds: 60), BackoffType.exponential),
          const Duration(seconds: 4),
        );
        expect(
          calculateBackoff(3, const Duration(seconds: 1),
              const Duration(seconds: 60), BackoffType.exponential),
          const Duration(seconds: 8),
        );
      });

      test('最大遅延を超えない', () {
        expect(
          calculateBackoff(20, const Duration(seconds: 1),
              const Duration(seconds: 30), BackoffType.exponential),
          const Duration(seconds: 30),
        );
      });
    });
  });

  group('addJitter', () {
    test('元の値の 75%~125% の範囲内に収まる', () {
      const delay = Duration(seconds: 10);
      for (var i = 0; i < 100; i++) {
        final result = addJitter(delay);
        expect(result.inMilliseconds, greaterThanOrEqualTo(7500));
        expect(result.inMilliseconds, lessThanOrEqualTo(12500));
      }
    });
  });
}

File: framework/frontend/react/packages/k1s0-realtime/tests/websocket/backoff.test.ts
/**
 * バックオフ計算のテスト
 */

import { describe, expect, it } from 'vitest';
import { calculateBackoff, addJitter } from '../../src/utils/backoff.js';

describe('calculateBackoff', () => {
  describe('linear backoff', () => {
    it('試行回数に比例して遅延が増加する', () => {
      expect(calculateBackoff(0, 1000, 30000, 'linear')).toBe(1000);
      expect(calculateBackoff(1, 1000, 30000, 'linear')).toBe(2000);
      expect(calculateBackoff(2, 1000, 30000, 'linear')).toBe(3000);
      expect(calculateBackoff(4, 1000, 30000, 'linear')).toBe(5000);
    });

    it('最大遅延を超えない', () => {
      expect(calculateBackoff(100, 1000, 30000, 'linear')).toBe(30000);
    });
  });

  describe('exponential backoff', () => {
    it('試行回数の指数関数で遅延が増加する', () => {
      expect(calculateBackoff(0, 1000, 60000, 'exponential')).toBe(1000);
      expect(calculateBackoff(1, 1000, 60000, 'exponential')).toBe(2000);
      expect(calculateBackoff(2, 1000, 60000, 'exponential')).toBe(4000);
      expect(calculateBackoff(3, 1000, 60000, 'exponential')).toBe(8000);
    });

    it('最大遅延を超えない', () => {
      expect(calculateBackoff(20, 1000, 30000, 'exponential')).toBe(30000);
    });
  });
});

describe('addJitter', () => {
  it('元の値の 75%~125% の範囲内に収まる', () => {
    const delay = 1000;
    for (let i = 0; i < 100; i++) {
      const result = addJitter(delay);
      expect(result).toBeGreaterThanOrEqual(750);
      expect(result).toBeLessThanOrEqual(1250);
    }
  });

  it('整数値を返す', () => {
    const result = addJitter(1000);
    expect(Number.isInteger(result)).toBe(true);
  });
});
